### PR TITLE
Final endpoints

### LINKED ISSUE #52: Wrong error: Get hackathons
I tried to GET `/hacks/`, but it gives me a `404` error. This is caused by the DAL which says there are no rows to show, but this probably shouldn't be a `404`. The correct response is an empty array `[]` which should be loaded into the application to show that no hackathons have been created.

### LINKED ISSUE #73: Error: Get registration questions
SQL syntax error. Should say `SELECT * FROM reg_options ...`

```
hackapp-db          | 2020-04-01 13:53:56.983 UTC [190] ERROR:  syntax error at or near "reg_options" at character 10
hackapp-db          | 2020-04-01 13:53:56.983 UTC [190] STATEMENT:  SELECT * reg_options WHERE qid = $1
hackapp-api         | { error: syntax error at or near "reg_options"
hackapp-api         |     at Connection.parseE (/usr/src/app/node_modules/pg/lib/connection.js:614:13)
hackapp-api         |     at Connection.parseMessage (/usr/src/app/node_modules/pg/lib/connection.js:413:19)
hackapp-api         |     at Socket.<anonymous> (/usr/src/app/node_modules/pg/lib/connection.js:129:22)
hackapp-api         |     at emitOne (events.js:116:13)
hackapp-api         |     at Socket.emit (events.js:211:7)
hackapp-api         |     at addChunk (_stream_readable.js:263:12)
hackapp-api         |     at readableAddChunk (_stream_readable.js:250:11)
hackapp-api         |     at Socket.Readable.push (_stream_readable.js:208:10)
hackapp-api         |     at TCP.onread (net.js:601:20)
hackapp-api         |   name: 'error',
hackapp-api         |   length: 100,
hackapp-api         |   severity: 'ERROR',
hackapp-api         |   code: '42601',
hackapp-api         |   detail: undefined,
hackapp-api         |   hint: undefined,
hackapp-api         |   position: '10',
hackapp-api         |   internalPosition: undefined,
hackapp-api         |   internalQuery: undefined,
hackapp-api         |   where: undefined,
hackapp-api         |   schema: undefined,
hackapp-api         |   table: undefined,
hackapp-api         |   column: undefined,
hackapp-api         |   dataType: undefined,
hackapp-api         |   constraint: undefined,
hackapp-api         |   file: 'scan.l',
hackapp-api         |   line: '1150',
hackapp-api         |   routine: 'scanner_yyerror' }

```

### PR DESCRIPTION
**Pull request by evscott**
**Main changes include:**

1. Changes to endpoint paths to better accommodate query parameters and prevent path collisions
2. Endpoint convention changed to use `GET` and `DELETE` with and `POST` and `PUT` without query parameters
3. Create registration question -> create registration _questions_
4. Implementation of get registration CSV endpoint
5. Various minor changes

### CODE REVIEW

**Comment by evscott** [comment]
The major endpoints now are:

### Authentication
- can sign up
- can login

### User
- can update a user
- can delete a user
- can get a user

### Hackathon
- can delete a hackathon and all of its related data

### Overview
- can create an overview
- can update an overview
- can get an overview for a hackathon
- can get an overview for every hackathon

### Details
- can create many details for a hackathon
- can update a detail for a hackathon
- can delete a delete for a hackathon
- can get details for a hackathon

### Registration questions
- can create many registration questions w/ options
- can update a registration question
- can delete a registration question
- can get all registration questions for a hackathon

### Registration question options
- can create a registration question option
- can update a registration question option
- can delete a registration question option

### Registration answers
- can create a registration question answer
- can update a registration question answer
- can get a string representing a CSV file that contains every answer to every registration questions associated with a hackathon
- can get a users answers to every registration questions associated with a hackathon

File: migrator/migrations/sqls/20200220183445-seed-down.sql [hunk 1]
```diff
@@ -1,11 +1,8 @@
-DROP TABLE IF EXISTS File_links;
```
Comment (gzinck) [inline migrator/migrations/sqls/20200220183445-seed-down.sql:1]: Haha shucks—wish we had the time to implement images!

File: src/dal/auth.js [hunk 2]
```diff
@@ -2,11 +2,10 @@ const pool = require('./dal-pool');
 
 async function signUp(firstName, lastName, email, password) {
     try {
-        let res = await pool.query('INSERT INTO users (first_name, last_name, email, password) VALUES ($1, $2, $3, $4) RETURNING uid, first_name, last_name, email, admin',
+        let res = await pool.query('INSERT INTO users (first_name, last_name, email, password) VALUES ($1, $2, $3, $4) RETURNING uid, first_name as "firstName", last_name as "lastName", email, admin',
```
Comment (gzinck) [inline src/dal/auth.js:5]: Cool—didn't know you could use quotes "firstName" like that!

File: src/dal/hack.js [hunk 3]
```diff
@@ -13,36 +61,138 @@ async function getHackathon(hid) {
     }
 }
 
-async function getHackathons() {
+async function getHackathonOverviews() {
```
Comment (gzinck) [inline src/dal/hack.js:64]: I like the new naming convention, it makes things much clearer

File: src/dal/reg.js [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+                    if (res.rowCount === 0) {
+                        return {question: null, options: null, err: 400};
+                    }
+
+                    optionsCreated.push(res.rows[0]);
+                }
+            }
+        }
+
+        return {questions: questionsCreated, options: optionsCreated, err: null}
```
Comment (gzinck) [inline src/dal/reg.js:44]: This must've been a challenging endpoint :) Looks good!

File: src/dal/reg.js [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+        console.error(err)
+        return {err: 500}
+    }
+}
+
+async function getRegAnswers(hid) {
+    try {
+        let res = await pool.query('SELECT question, COALESCE(o.option, a.answer) as answer, u.first_name as "firstName", u.last_name as "lastName", u.email ' +
+                                   'FROM reg_questions AS q JOIN reg_answers AS a ON q.qid = a.qid FULL JOIN reg_options as o ON o.oid = a.oid FULL JOIN users as u ON u.uid = a.uid ' +
+                                   'WHERE a.oid IS NOT NULL OR a.answer IS NOT NULL AND q.hid = $1',
```
Comment (gzinck) [inline src/dal/reg.js:266]: OMG 👏👏👏

File: src/handlers/reg.js [hunk 6]
```diff
@@ -1,94 +1,931 @@
...
+ *                    oid:
+ *                      type: string
+ *                      format: uuid
+ *                    answer:
+ *                      type: string
+ *              answersDeleted:
+ *                type: array
+ *                items:
+ *                  type: string
+ *                  format: uuid
```
Comment (gzinck) [inline src/handlers/reg.js:859]: VERBOSE. WOW. But hey, it makes gorgeous Swagger docs for the API, so I support this. Having the return type for the response is really helpful.

File: src/handlers/reg.js [hunk 7]
```diff
@@ -106,13 +948,36 @@ let updateRegQuestion = async (req, res) => {
...
+    });
+    let csv = csvStringifier.getHeaderString() + csvStringifier.stringifyRecords(getRegAnswersRes.answers);
+
+    return res.status(200).send({csv: csv})
 };
 
 /**
  * @swagger
- * /hacks/:hid/reg/users/csv/:
+ * /u/hacks/reg/users/csv/:hid/:
```
Comment (gzinck) [inline src/handlers/reg.js:980]: Is this still the endpoint used, or has this changed?

File: test/hack.test.js [hunk 8]
```diff
@@ -22,251 +22,673 @@ describe('hackathon', () => {
...
+                        done();
+                    })
+            });
+
+            it('creating hackathon details with missing token should fail', function(done) {
+                let details = [
+                    {
+                        detail: 'detail one',
+                        index: 0,
+
```
Comment (gzinck) [inline test/hack.test.js:381]: Weird whitespace here, but not a big deal

**Review by gzinck — APPROVED** [review]
Most of my review was in the form of issues, so after reading through the code, nothing to add! Looks good man, lots of testing 👏👏👏

**Pull request merged**